netutils/netinit/netinit.c

```diff
@@ -343,6 +353,86 @@ static void netinit_set_macaddr(void)
 #  define netinit_set_macaddr()
 #endif
 
+#ifdef CONFIG_FSUTILS_IPCFG
+/****************************************************************************
+ * Name: netinit_get_ipaddrs
+ *
+ * Description:
+ *   Setup IP addresses.
+ *
+ *   For 6LoWPAN, the IP address derives from the MAC address.  Setting it
+ *   to any user provided value is asking for trouble.
+ *
+ ****************************************************************************/
+
+static int netinit_get_ipaddrs(void)
+{
+#ifdef CONFIG_NETINIT_DHCPC
+  bool use_dhcp   = false;
+#endif
+  bool use_static = false;
+#ifdef CONFIG_NET_IPv4
+  struct in_addr addr;
+  int ret;
+
+  ret = ipcfg_read(NET_DEVNAME, (FAR struct ipcfg_s *) &g_netconf, AF_INET);
+  if (ret < 0)
+    {
+      return ret;
+    }
+  else
+    {
+      use_static = g_netconf.proto & BOOTPROTO_STATIC;
+
+#ifdef CONFIG_NETINIT_DHCPC
+      use_dhcp = g_netconf.proto & IPv4PROTO_DHCP;
+      addr.s_addr = use_dhcp ? 0 : HTONL(g_netconf.ipaddr);
+#else
```

Comment:
g_netconf.ipaddr is already in network order.  HTONL should not be used.

Reply:
Since the original code has the #defined values in host order.

https://github.com/nuttx-to-asf/incubator-nuttx-apps/blob/5a6246c3f2adfd0a4cc10639eb9923682e8ce1f4/netutils/netinit/netinit.c#L472-L477

 I choose that convention.  Values in code and memory on the host are in host order. My application does a hton on load (after the read) and ntol on save (before the save). but avoid all that would be good.  

For the PX4 use case this was necessary for future exposing the configuration data in QCG.

Reply:
> 
> 
> Since the original code has the #defined values in host order.
> 
> https://github.com/nuttx-to-asf/incubator-nuttx-apps/blob/5a6246c3f2adfd0a4cc10639eb9923682e8ce1f4/netutils/netinit/netinit.c#L472-L477
> 
> I choose that convention. Values in code and memory on the host are in host order. My application does a hton on load (after the read) and ntoh on save (before the save). but avoid all that would be good.
> 
> For the PX4 use case this was necessary for future exposing the configuration data in QCG.

Then you should use NTOHL not HTONL.  It would be a better design to use all network order internally.  Of course the user provided addresses from the Kconfigs will be in host order but applying HTONL in that case has zero overhead; the endian-ness is resolved at compile time.  Converting variable data to host order only so that it can be converted back to network order is very inefficient.

It is also expensive because NTOHL and HTON is relatively big when they operate on variable data (as well as non-standard.)  The better (and standard) interfaces might be ntohl() and htonl() in these cases.  Best of all:  Keep everything in network order.

Reply:
Well, looking at the code again, it is wrong.  The s_addr must always be in network order.  You save it here in host order which is inconsistent with all other uses and would fail.  All other settings in the original:

    netinit.c:  addr.s_addr = HTONL(CONFIG_NETINIT_IPADDR);
    netinit.c:  addr.s_addr = 0;
    netinit.c:  addr.s_addr = HTONL(CONFIG_NETINIT_DRIPADDR);
    netinit.c:  addr.s_addr = HTONL(CONFIG_NETINIT_NETMASK);
    netinit.c:  addr.s_addr = HTONL(CONFIG_NETINIT_DNSIPADDR);

The configuration settings are ALL in host order and the HTONL converts them to network order which is correct.  But this is incorrect:

    addr.s_addr = use_dhcp ? 0 : HTONL(g_netconf.ipaddr);

Because g_netconfig.ipaddr is in network order.  The HTONL is incorrect, it should be NTOHL.  But that is still wong because s_addre must be in network order.  This fix is to remove the HTONL altogether.